File: backend/src/main/java/com/astroluxeon/jltgbfv/service/GameService.java
package com.astroluxeon.jltgbfv.service;

import com.astroluxeon.jltgbfv.model.*;
import org.springframework.stereotype.Service;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;
import java.io.IOException;
import java.util.*;


@Service
public class GameService {
    private final int ACTIVE_DECK_SIZE = 6;
    private final String DATA_FILE = "game-data.json";
    private final String DEFAULT_FILE = "game-default.json";
    private final ObjectMapper objectMapper = new ObjectMapper();

    private Map<String, Region> regions;
    private List<Challenge> challenges;
    private List<Challenge> battles;
    private Stack<Pair<Integer, Integer>> challengeHistory;

    public GameService() {
        loadFromFile();
    }

    public List<Challenge> getChallenges() {
        return challenges;
    }

    public List<Challenge> getBattles() {
        return battles;
    }

    public Collection<Region> getRegions() {
        return regions.values();
    }

    private Challenge drawCard(List<Challenge> deck) {
        List<Challenge> cards = new ArrayList<>();

        for (Challenge c : deck) {
            if (c.getStatus() == Status.INACTIVE) {
                cards.add(c);
            }
        }

        if (!cards.isEmpty()) {
            Collections.shuffle(cards);
            Challenge c = cards.get(0);
            c.setStatus(Status.ACTIVE);
            return c;
        }

        return null;
    }

    public synchronized void completeChallenge(int id, String team) {
        Challenge challenge = null;

        for (Challenge c : challenges) {
            if (c.getId() == id) {
                challenge = c;
                break;
            }
        }

        if (challenge != null && challenge.getStatus() == Status.ACTIVE) {
            challenge.setStatus(Status.COMPLETED);
            challenge.setTeam(Team.valueOf(team));

            Challenge c = drawCard(challenges);
            int id2 = c != null ? c.getId() : -1;
            challengeHistory.push(new Pair<>(id, id2));
            saveToFile();
        }
    }

    public Challenge startBattle() {
        Challenge c = drawCard(battles);
        saveToFile();
        return c;
    }

    public synchronized void completeBattle(int id, String team) {
        Challenge battle = null;

        for (Challenge c : battles) {
            if (c.getId() == id) {
                battle = c;
                break;
            }
        }

        if (battle != null && battle.getStatus() == Status.ACTIVE) {
            battle.setStatus(Status.COMPLETED);
            battle.setTeam(Team.valueOf(team));
            saveToFile();
        }
    }

    public void claimRegion(String id, String team) {
        if (regions.get(id) != null) {
            try {
                regions.get(id).setTeam(Team.valueOf(team));
                saveToFile();
                System.out.println(regions.get(id).getName() + " claimed by " + team);
            } catch (Exception e) {
                System.out.println(team + " is not a valid team.");
            }
        }
    }

    public void lockRegion(String id, boolean lock) {
        if (regions.get(id) != null) {
            regions.get(id).setLocked(lock);
            saveToFile();
            if (lock) {
                System.out.println(regions.get(id).getName() + " locked by " + regions.get(id).getTeam());
            }
        }
    }

    public void undoChallenge() {
        if (challengeHistory.isEmpty()) {
            return;
        }

        Pair<Integer, Integer> top = challengeHistory.pop();
        int id1 = top.key();
        int id2 = top.value();

        for (Challenge challenge : challenges) {
            if (challenge.getId() == id1) {
                challenge.setStatus(Status.ACTIVE);
                challenge.setTeam(Team.NONE);
                break;
            }
        }

        if (id2 != -1) {
            for (Challenge challenge : challenges) {
                if (challenge.getId() == id2) {
                    challenge.setStatus(Status.INACTIVE);
                    break;
                }
            }
        }

        saveToFile();
    }

    public void undoBattle(int id) {
        for (Challenge c : battles) {
            if (c.getId() == id) {
                if (c.getStatus() == Status.COMPLETED) {
                    c.setStatus(Status.ACTIVE);
                    c.setTeam(Team.NONE);
                } else if (c.getStatus() == Status.ACTIVE) {
                    c.setStatus(Status.INACTIVE);
                    c.setTeam(Team.NONE);
                }
                saveToFile();
                break;
            }
        }
    }

    public void initializeNewGame() {
        File file = new File(DEFAULT_FILE);
        if (file.exists()) {
            try {
                GameState state = objectMapper.readValue(file, GameState.class);
                this.regions = state.getRegions();
                this.challenges = state.getChallenges();
                this.battles = state.getBattles();
                this.challengeHistory = state.getChallengeHistory();
            } catch (IOException e) {
                System.out.println("Error initializing game: " + e.getMessage());
                this.regions = new HashMap<>();
                this.challenges = new ArrayList<>();
                this.battles = new ArrayList<>();
                this.challengeHistory = new Stack<>();
            }
        } else {
            System.out.println("No data found.");
            this.regions = new HashMap<>();
            this.challenges = new ArrayList<>();
            this.battles = new ArrayList<>();
            this.challengeHistory = new Stack<>();
        }

        for (int i = 0; i < ACTIVE_DECK_SIZE; i++) {
            drawCard(challenges);
        }

        System.out.println("Initialized new game.");
        saveToFile();
    }

    private void loadFromFile() {
        File file = new File(DATA_FILE);
        if (file.exists()) {
            try {
                GameState state = objectMapper.readValue(file, GameState.class);
                this.regions = state.getRegions();
                this.challenges = state.getChallenges();
                this.battles = state.getBattles();
                this.challengeHistory = state.getChallengeHistory() != null ? state.getChallengeHistory() : new Stack<>();
                System.out.println("Loaded save data.");
            } catch (IOException e) {
                System.out.println("Error reading save data: " + e.getMessage());
                initializeNewGame();
            }
        } else {
            System.out.println("No save data found.");
            initializeNewGame();
        }
    }

    private void saveToFile() {
        try {
            GameState state = new GameState(this.regions, this.challenges, this.battles, this.challengeHistory);
            objectMapper.writeValue(new File(DATA_FILE), state);
        } catch (IOException e) {
            System.out.println("Error saving file: " + e.getMessage());
        }
    }
}

File: backend-node/src/game-service.ts
import fs from 'fs/promises';
import path from 'path';
import { Region, Challenge, Team } from './types';

export interface GameData {
  regions: Record<string, Region>;
  challenges: Challenge[];
  battles: Challenge[];
  challengeHistory: [number, number][];
}

const DATA_FILE = path.join(process.cwd(), 'game-data.json');
const DEFAULT_FILE = path.join(process.cwd(), 'game-default.json');
const ACTIVE_DECK_SIZE = 6;

function shuffle<T>(array: T[]): void {
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j]!, array[i]!];
  }
}

class GameService {
  private data: GameData = {
    regions: {},
    challenges: [],
    battles: [],
    challengeHistory: []
  };

  public async init(): Promise<void> {
    await this.loadFromFile();
  }

  private async loadFromFile(): Promise<void> {
    try {
      const data = await fs.readFile(DATA_FILE, 'utf-8');
      this.data = JSON.parse(data);
      console.log('Loaded save data.');
    } catch (error) {
      console.log('No save data found or error reading data. Initializing new game...');
      await this.initializeNewGame();
    }
  }

  private async saveToFile(): Promise<void> {
    try {
      const data = JSON.stringify(this.data, null, 2);
      await fs.writeFile(DATA_FILE, data, 'utf-8');
    } catch (error) {
      console.error('Error saving file:', error);
    }
  }

  public async initializeNewGame(): Promise<void> {
    try {
      const data = await fs.readFile(DEFAULT_FILE, 'utf-8');
      this.data = JSON.parse(data);
      this.data.challengeHistory = [];
    } catch (error) {
      console.error('Error initializing default game data:', error);
      this.data = { regions: {}, challenges: [], battles: [], challengeHistory: [] };
    }

    for (let i = 0; i < ACTIVE_DECK_SIZE; i++) {
      this.drawCard(this.data.challenges);
    }

    console.log('Initialized new game.');
    await this.saveToFile();
  }

  public getChallenges(): Challenge[] {
    return this.data.challenges;
  }

  public getBattles(): Challenge[] {
    return this.data.battles;
  }

  public getRegions(): Region[] {
    return Object.values(this.data.regions);
  }

  private drawCard(deck: Challenge[]): Challenge | null {
    const inactiveCards = deck.filter(c => c.status === 'INACTIVE');

    if (inactiveCards.length > 0) {
      shuffle(inactiveCards);
      const challenge = inactiveCards[0];

      const target = deck.find(c => c.id === challenge?.id);
      if (target) {
        target.status = 'ACTIVE';
        return target;
      }
    }
    return null;
  }

  public async completeChallenge(id: number, team: Team): Promise<void> {
    const challenge = this.data.challenges.find(c => c.id === id);

    if (challenge && challenge.status === 'ACTIVE') {
      challenge.status = 'COMPLETED';
      challenge.team = team;

      const newCard = this.drawCard(this.data.challenges);
      const id2 = newCard ? newCard.id : -1;

      this.data.challengeHistory.push([id, id2]);
      await this.saveToFile();
    }
  }

  public async startBattle(): Promise<Challenge | null> {
    const c = this.drawCard(this.data.battles);
    await this.saveToFile();
    return c;
  }

  public async completeBattle(id: number, team: Team): Promise<void> {
    const battle = this.data.battles.find(c => c.id === id);

    if (battle && battle.status === 'ACTIVE') {
      battle.status = 'COMPLETED';
      battle.team = team;
      await this.saveToFile();
    }
  }

  public async claimRegion(id: string, team: Team): Promise<void> {
    const region = this.data.regions[id];
    if (region) {
      region.team = team;
      await this.saveToFile();
      console.log(`${region.name} claimed by ${team}`);
    } else {
      console.log(`Region ${id} not found.`);
    }
  }

  public async lockRegion(id: string, lock: boolean): Promise<void> {
    const region = this.data.regions[id];
    if (region) {
      region.locked = lock;
      await this.saveToFile();
      if (lock) {
        console.log(`${region.name} locked by ${region.team}`);
      }
    }
  }

  public async undoChallenge(): Promise<void> {
    if (this.data.challengeHistory.length === 0) return;

    const [id1, id2] = this.data.challengeHistory.pop()!;

    const challenge1 = this.data.challenges.find(c => c.id === id1);
    if (challenge1) {
      challenge1.status = 'ACTIVE';
      challenge1.team = 'NONE';
    }

    if (id2 !== -1) {
      const challenge2 = this.data.challenges.find(c => c.id === id2);
      if (challenge2) {
        challenge2.status = 'INACTIVE';
      }
    }

    await this.saveToFile();
  }

  public async undoBattle(id: number): Promise<void> {
    const battle = this.data.battles.find(c => c.id === id);

    if (battle) {
      if (battle.status === 'COMPLETED') {
        battle.status = 'ACTIVE';
        battle.team = 'NONE';
      } else if (battle.status === 'ACTIVE') {
        battle.status = 'INACTIVE';
        battle.team = 'NONE';
      }
      await this.saveToFile();
    }
  }
}

export const gameService = new GameService();